File: LeetCode/JavaScript/#20 Valid Parenttheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    var all = {"(":")","{":"}","[":"]"};
    var res = [];
    var len = s.length;
    for(let i =0;i<len;i++){
      if(s[i] in all){
        res.push(all[s[i]]);
      }else{
        if(s[i]!==res.pop()){
          return false;
        }
      }
    }
    return res.length === 0;
};

File: LeetCode/Python/#20 Valid Parenttheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        alls = {")":"(","}":"{","]":"["}
        res = []

        for i in s:
            if i in alls.values():
                res.append(i)
            else:
                if not res or res.pop()!=alls[i]:
                    return False

        return len(res)==0


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        lens = len(s)
        while(lens!=0 and lens%2==0):
            s = s.replace("()","")
            s = s.replace("[]","")
            s = s.replace("{}","")

            lens = len(s) if lens>len(s) else 0

        return len(s)==0
